src/urllib3/_collections.py

```diff
@@ -220,6 +220,8 @@ def add(self, key, val):
         >>> headers['foo']
         'bar, baz'
         """
+        if PY3 and isinstance(val, bytes):
+            val = val.decode("utf-8")
```

Comment:
I don't know that we unconditionally want to force things to utf8 as that may not always be safe to do for data validity/consistency

Reply:
Thanks for review. I changed it to `val = val.decode()`.

Reply:
Note that this strictly equivalent since in Python 3 the default encoding is 'utf-8': https://docs.python.org/3/library/stdtypes.html#bytes.decode. (By the way, we don't want to use `if PY3`, see https://github.com/urllib3/urllib3/pull/1669)

Honestly, I don't know what the real fix should be, but hey @sethmlarson is often optimistic when adding "Contributor Friendly" labels. :D

Reply:
It would helpful to first have a way to reproduce the problem, as it would help us to understand it better. Using boto3 would be fine, but it shouldn't require AWS credentials. @hramezani Do you think you can try to work on a such a reproducer script?

Reply:
What are our thoughts on adding a `ensure_str(x, "ascii")` to each header value within `HTTPHeaderDict.__getitem__()`?
The `__getitem__()` call would previously explode if users were adding bytes anyways so we know we're not "breaking" anyone and it seems like handing a `bytes` instance to httplib hasn't caused problems in the past.
Going that route preserves the ability to have `bytes` header types in `HTTPHeaderDict` which we've been unknowingly supporting all this time.

Reply:
`ensure_str` encodes strings to bytes? It's the only safe operation, yes, and `ascii` is the encoding used by httplib anyway, so that would make a lot of sense indeed. 👍 on that

Reply:
Thanks @sethmlarson for your suggestion. I changed my PR based on that. Don't know I got it right or not.

> Do you think you can try to work on a such a reproducer script?

@pquentin
Here is a simple script from botocore GitHub page:
```python
>>> import botocore.session
>>> session = botocore.session.get_session()
>>> client = session.create_client('ec2')
>>> print(client.describe_instances())
```
I reproduced the bug with the above script in my machine. Here are the steps:

- Create a virtualenv
- Install botocore `pip install botocore`
- Remove urllib3 `pip uninstall urllib3`
- Install urllib3 from my local repository `pip install -e <urllib3 path on my machine>`
- Run the script - Here I got the error `botocore.exceptions.HTTPClientError: An HTTP Client raised an unhandled exception: sequence item 0: expected str instance, bytes found`
- Apply the patch
- Run the script - Was successful